File: swift/imageio.swift
// Module of functions for writing PGM and PBM files in Swift.
// Owain Kenway

import Foundation

// A private class to make writing to OutputStream less of a mess.
// (i.e. more like other programming langauges!)
fileprivate class textfile {
    var filename : String
    var stream : OutputStream

    init(filename: String) {
        self.filename = filename
        self.stream = OutputStream(toFileAtPath: filename, append: false)!
    }

    func open() {
        self.stream.open()
    }

    func close() {
        self.stream.close()
    }

// This is the really unpleasant one method (i.e. why we are doing this.)
    func write(data: String) {
        let retval: Int = self.stream.write(data, maxLength: data.lengthOfBytes(using: String.Encoding.utf8))
        if retval == -1 {
            print("Something has gone horribly wrong.")
            exit(-1)
        }
    }
}

// Method to write 2D array d as a PGM image.
// white: highest value in d
// filename: file to write to
public func writepgm(d: [[Int]], white: Int, filename: String) {
    let x: Int = d.count
    let y: Int = d[0].count
 
 // Open output stream.
    let out: textfile = textfile(filename: filename)
    out.open()

// Write header.
    out.write(data: "P2\n")
    out.write(data: "# Written by pnmmodules (https://github.com/owainkenwayucl/pnmmodules).\n")
    out.write(data: String(x) + " " + String(y) + "\n")
    out.write(data: String(white) + "\n")

    for j in 1...y {
        for i in 1...x {
            out.write(data: String(d[i-1][j-1]) + "\n")
        }
    }
    out.close()

} 

// Method to write 2D array d as a PBM image.
// threshold: value at which a pixel becomes 1.
// filename: file to write to
public func writepbm(d: [[Int]], threshold: Double, filename: String) {
    let x: Int = d.count
    let y: Int = d[0].count
 
 // Open output stream.
    let out: textfile = textfile(filename: filename)
    out.open()

// Write header.
    out.write(data: "P1\n")
    out.write(data: "# Written by pnmmodules (https://github.com/owainkenwayucl/pnmmodules).\n")
    out.write(data: String(x) + " " + String(y) + "\n")

    for j in 1...y {
        for i in 1...x {
            if (Double(d[i-1][j-1]) >= threshold) {
                out.write(data: "1\n")
            } else {
                out.write(data: "0\n")
            }
        }
    }
    out.close()

} 

File: c/src/imageio.c
#include <stdio.h>

/*
 Write a pgm file from array at pointer d:
  xres     : x resolution
  yres     : y resolution
  white    : max value (corresponds to white)
  filename : file to write 
*/
void writepgm(int** d, int xres, int yres, int white, char filename[]) {
    int i, j;
    FILE *outfile;

// Open file.
    outfile = fopen(filename, "w");

// Write our header.
    fprintf(outfile, "P2\n");
    fprintf(outfile, "# Written by pnmmodules (https://github.com/owainkenwayucl/pnmmodules).\n");
    fprintf(outfile, "%d %d\n%d\n", xres, yres, white);

// Write out the array.
    for (j = 0; j < yres; j++){
        for (i = 0; i < xres; i++) {
            fprintf(outfile, "%d\n", d[i][j]);
        }
    }

// Close the file.
    fclose(outfile);
}

/*
 Write a pbm file from array at pointer d:
  xres      : x resolution
  yres      : y resolution
  threshold : value at which pixel becomes equal to 1.
  filename  : file to write 
*/
void writepbm(int** d, int xres, int yres, int threshold, char filename[]) {
    int i, j;
    FILE *outfile;

// Open file.
    outfile = fopen(filename, "w");

// Write our header.
    fprintf(outfile, "P1\n");
    fprintf(outfile, "# Written by pnmmodules (https://github.com/owainkenwayucl/pnmmodules).\n");
    fprintf(outfile, "%d %d\n", xres, yres);

// Write out the array.
    for (j = 0; j < yres; j++){
        for (i = 0; i < xres; i++) {
            if (d[i][j] >= threshold) {
                fprintf(outfile, "1\n");
            } else {
                fprintf(outfile, "0\n");
            }
        }
    }

// Close the file.
    fclose(outfile);
}